### DISCUSSION TITLE
zh/datasets/obb/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#16923] [Docs]
# zh/datasets/obb/

了解用于Ultralytics YOLO  模型的 OBB 数据集格式。了解它们的结构、应用和格式转换，以增强物体检测训练。

https://docs.ultralytics.com/zh/datasets/obb/

**Comment by QLGalaxy** [comment]
Hi！I have transformed my data format from DOTA to yolo after spliting to 1024x1024. And I want to predict on the 1024*1024 test imgs, combine them and convert the format for submission to the official website for testing. Is there any tools for this?

**Reply by glenn-jocher** [reply]
@QLGalaxy to combine predictions and convert them for submission, you can use post-processing tools to merge the 1024x1024 tiles back into the original image format. While Ultralytics doesn't provide a specific tool for this, you might consider using image processing libraries like OpenCV or PIL to achieve the desired output. For more details on dataset handling, visit our documentation at https://docs.ultralytics.com/datasets/.

**Reply by QLGalaxy** [reply]
Thanks for your explaination. And further, I wonder how many GPUs do you use on training yolov11n-OBB.pt？

**Reply by glenn-jocher** [reply]
The number of GPUs used for training YOLO11n-OBB.pt can vary based on resources and requirements. Typically, multi-GPU setups are used to speed up training, but a single GPU can also suffice depending on the dataset size and model complexity. For specific configurations, you might want to experiment based on your hardware capabilities.

**Comment by UltralyticsAssistant** [comment]
👋 Hello, thank you for reaching out and showing interest in Ultralytics 🚀! Our [Documentation](https://docs.ultralytics.com) is a great place to explore for detailed information on datasets and formats, including the OBB dataset format you're interested in. Be sure to check out the [OBB section](https://docs.ultralytics.com/zh/datasets/obb/) for insights on their structure, application, and format conversion.

If this is a 🐛 Bug Report related to datasets, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum_reproducible_example/) to help us investigate further.

For any custom training ❓ Questions, including dataset details or model-specific queries, please provide additional context such as dataset samples and training logs. Also, consider reviewing our [Training Tips](https://docs.ultralytics.com/guides/model-training-tips/) for best practices.

Join the Ultralytics community for more interactive discussions. You can chat with us in real-time on [Discord](https://ultralytics.com/discord) 🎧, explore in-depth topics on [Discourse](https://community.ultralytics.com), or engage with others on our [Subreddit](https://reddit.com/r/ultralytics).

## Upgrade

Ensure you have the latest `ultralytics` package and all its [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) installed in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in various environments with pre-installed dependencies:

- **Notebooks** with GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov8"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud**: See [GCP Quickstart](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **AWS**: See [AWS Quickstart](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker**: See [Docker Quickstart](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yaml/badge.svg" alt="Ultralytics CI"></a>

This badge indicates the status of our Continuous Integration (CI) tests, ensuring the correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) across different platforms.

This is an automated response, but an Ultralytics engineer will be with you soon to provide further assistance. 😊

**Comment by LinkZY3471** [comment]
I encountered an issue while training on my custom dataset. In my dataset, some of the Oriented Bounding Boxes (OBB) are large and rotated, causing some corner coordinates to exceed the original image boundaries. This leads to values greater than 1 or less than 0 when converting to YOLO format, resulting in these samples being discarded during YOLO model training.
How should I handle these OBB boxes that exceed image boundaries? What's the recommended approach to process these cases while maintaining the integrity of the training data?

**Reply by glenn-jocher** [reply]
@LinkZY3471 to handle Oriented Bounding Boxes (OBBs) that exceed image boundaries, clip the corner coordinates to the valid [0, 1] range while ensuring the bounding box integrity. This adjustment prevents the loss of samples during training. For further details on managing OBB datasets, refer to the [Ultralytics OBB guide](https://docs.ultralytics.com/datasets/obb/).

**Reply by LinkZY3471** [reply]
Thank you, but if we simply clip the out-of-range values to 0-1, can the OBB shape still be correctly maintained? Would this cause any problems during training?

**Reply by glenn-jocher** [reply]
@LinkZY3471 clipping OBB coordinates to the [0, 1] range may slightly distort the box shape, especially if significant portions extend beyond the image boundary. However, this ensures all annotations remain valid for training. If preserving the exact OBB shape is critical, you could preprocess the data to crop or adjust annotations while maintaining their integrity. For details on handling OBB datasets, refer to [Ultralytics OBB documentation](https://docs.ultralytics.com/datasets/obb/).

**Reply by LinkZY3471** [reply]
Thank you :)

**Reply by glenn-jocher** [reply]
@LinkZY3471 you're welcome! If clipping affects OBB integrity significantly, consider preprocessing to adjust or split overly large boxes to fit within image bounds while maintaining their shape. For further insights, refer to the [Ultralytics OBB guide](https://docs.ultralytics.com/datasets/obb/). Let us know if you need additional clarification!

**Comment by luckg6** [comment]
Hello Ultralytics team,

I have a question regarding YOLO OBB (Oriented Bounding Box) object detection. I noticed that for predicted bounding boxes, the angle can be obtained via obb.xywhr. However, it seems that there is no direct way to get the angle of the ground-truth boxes in the dataset.

Does this mean that if I want to calculate the angle error, I must manually compute it by converting the 8-point coordinates of the ground-truth boxes into an angle?

Thank you for your clarification!

**Reply by glenn-jocher** [reply]
GT labels in YOLO OBB are stored as 8 normalized corner coords (`class x1 y1 ... x4 y4`), so there’s no explicit angle field in the `*.txt`; if you want an angle/error metric, convert the GT corners to `xywhr` using the same helper Ultralytics uses internally via [`xyxyxyxy2xywhr`](https://docs.ultralytics.com/reference/utils/ops/) (angle returned in radians), which matches how `results[0].obb.xywhr` is defined in the [`OBB` results API](https://docs.ultralytics.com/reference/engine/results/) and follows the OBB conventions noted in the [OBB task docs](https://docs.ultralytics.com/tasks/obb/).

```python
import cv2
import math
import torch
from ultralytics.utils.ops import xyxyxyxy2xywhr

img = cv2.imread("path/to/image.png")
h, w = img.shape[:2]

# label line: cls x1 y1 x2 y2 x3 y3 x4 y4  (all coords normalized 0-1)
parts = open("path/to/label.txt").read().strip().split()
pts = torch.tensor([float(x) for x in parts[1:]], dtype=torch.float32).view(1, 4, 2)
pts[..., 0] *= w
pts[..., 1] *= h

gt_xywhr = xyxyxyxy2xywhr(pts.view(1, 8))
gt_theta = gt_xywhr[0, 4].item()  # radians

def angle_diff_rad(pred_theta, gt_theta):
    # smallest signed diff considering 90-degree periodicity used by OBB
    return (pred_theta - gt_theta + math.pi / 2) % math.pi - math.pi / 2
```

If you have follow-ups on the exact angle convention for your evaluation, drop them here in this discussion so the team can see the context.
